java/Cherepnin_Oleh/src/com/zoolatech/lecture4/tasks/_2/Tree.java

```diff
@@ -0,0 +1,88 @@
+package com.zoolatech.lecture4.tasks._2;
+
+public class Tree {
+    private Node head;
+
+    public Tree() {
+    }
+
+    public boolean add(int value) {
+        Node newNode = new Node(value);
+
+        if (head == null) {
+            head = newNode;
+            return true;
+        }
+
+        Node current = this.head;
+        do {
+            if (value == current.getValue()) {
+                return false;
+            }
+            if (value < current.getValue()) {
+                if (current.getLeft() == null) {
+                    current.setLeft(newNode);
+                    return true;
+                } else if (current.getLeft() != null) {
+                    current = current.getLeft();
+                    continue;
+                }
+            }
+            if (value > current.getValue()) {
+                if (current.getRight() == null) {
+                    current.setRight(newNode);
+                    return true;
+                } else if (current.getRight() != null) {
+                    current = current.getRight();
+                }
+            }
+        } while (true);
+    }
+
+    public void print() {
+        showNodes(head);
+    }
+
+    private void showNodes(Node startNode) {
+        if (startNode != null) {
+            showNodes(startNode.getLeft());
+            startNode.printNode();
+            showNodes(startNode.getRight());
+        }
+    }
+
+    private class Node {
```

Comment:
why have you defined this class as inner? does it need to have an access to some state of an outer class?